src/chains/ethereum/ethereum/src/helpers/simulation-handler.ts

```diff
@@ -0,0 +1,487 @@
+    this.#vm.on("step", (event: InterpreterStep) => {
+      const logs = maybeGetLogs(event);
+      if (logs) {
+        this.emit("ganache:vm:tx:console.log", {
+          context: this.#transactionContext,
+          logs
+        });
+      }
+      const ganacheStepEvent = makeStepEvent(this.#transactionContext, event);
+      this.emit("ganache:vm:tx:step", ganacheStepEvent);
+    });
+  };
+
+  #emitBefore = () => {
+    const context = (this.#transactionContext = {});
+    this.emit("ganache:vm:tx:before", {
+      context
+    });
+  };
+
+  #emitAfter = () => {
+    this.emit("ganache:vm:tx:after", {
+      context: this.#transactionContext
+    });
+  };
+
+  public toLightEJSAddress(address?: Address): EthereumJsAddress {
+    if (address) {
+      const buf = address.toBuffer();
+      return { buf, equals: (a: { buf: Buffer }) => buf.equals(a.buf) } as any;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Runs `vm.runCall` based off of the options set from running `initialize`.
+   * Throws if there is an `exceptionError` on the result.
+   * @returns The `EVMResult` from running the transaction.
+   */
+  public async runCall(): Promise<EVMResult> {
+    let callResult: EVMResult;
+    if (this.#initializationError) {
+      callResult = {
+        execResult: {
+          runState: { programCounter: 0 },
+          exceptionError: new VmError(this.#initializationError),
+          returnValue: BUFFER_EMPTY
+        }
+      } as any;
+    } else if (this.#runCallOpts) {
+      this.#emitBefore();
+      this.#setupStepEventEmits();
+      callResult = await this.#vm.runCall(this.#runCallOpts);
+      this.#emitAfter();
+    } else {
+      throw new Error("Unknown error when simulating transaction.");
+    }
+
+    if (callResult.execResult.exceptionError) {
+      throw new CallError(callResult);
+    } else {
+      return callResult;
+    }
+  }
+
+  /**
+   * Iteratively runs the `vm.runCall` function until the touched addresses and
+   * storage are the same as the previously generated access list.
+   * @param initialAccessList An access list to compare against the generated
+   * access list.
+   * @returns The final access list generated, plus an estimate of the gas
+   * consumed by running the transaction _with_ the generated access list included.
+   */
+  public async createAccessList(
+    initialAccessList?: AccessList
+  ): Promise<{ accessList: AccessList; gasUsed: string }> {
+    // no real reason why this is our max, feel free to change
+    const MAX_ITERATIONS = 1000;
+    let previousAccessList = initialAccessList || [];
+    const accessListExclusions = this.#accessListExclusions;
+    const addressesOnlyStorage = this.#addressesOnlyStorage;
+    let iterations = 0;
+    do {
+      // checkpoint so we can get back to this vm state after every time we
+      // run this loop
+      await this.#stateManager.checkpoint();
+      const stateManager = this.#stateManager;
```

Comment:
move `const stateManager = this.#stateManager;` outside the loop (and then don't use `this.#stateManager` again)

Reply:
Ooops, good catch!